### PR TITLE
Feature/osgng

### PR DESCRIPTION
**Pull request by marksantcroos**
Putting this up for incremental review.

### CODE REVIEW

File: src/radical/pilot/agent/radical-pilot-agent-multicore.py [hunk 1]
```diff
@@ -3242,7 +3247,8 @@ def stop(self):
         # During LRMS termination, we call any existing shutdown hooks on the
         # launch methods.  We only call LM shutdown hooks *once*
         launch_methods = set() # set keeps entries unique
-        launch_methods.add(self._cfg['mpi_launch_method'])
+        if 'mpi_launch_method' in self._cfg:
+            launch_methods.add(self._cfg['mpi_launch_method'])
         launch_methods.add(self._cfg['task_launch_method'])
```
Comment (andre-merzky) [inline src/radical/pilot/agent/radical-pilot-agent-multicore.py]: The task_lm would also need a guard?  Or we use `cfg.get()` and later ignore the `None` in the set?

File: src/radical/pilot/controller/pilot_launcher_worker.py [hunk 2]
```diff
@@ -362,16 +362,17 @@ def run(self):
...
+                        runtime         = compute_pilot['description']['runtime']
+                        queue           = compute_pilot['description']['queue']
+                        project         = compute_pilot['description']['project']
+                        cleanup         = compute_pilot['description']['cleanup']
+                        resource_key    = compute_pilot['description']['resource']
+                        schema          = compute_pilot['description']['access_schema']
+                        memory          = compute_pilot['description']['memory']
+                        candidate_hosts = compute_pilot['description']['candidate_hosts']
+                        pilot_sandbox   = compute_pilot['sandbox']
+                        global_sandbox  = compute_pilot['global_sandbox']
```
Comment (andre-merzky) [inline src/radical/pilot/controller/pilot_launcher_worker.py]: Thanks :)

File: src/radical/pilot/controller/pilot_launcher_worker.py [hunk 3]
```diff
@@ -663,28 +667,32 @@ def run(self):
...
                         msg = "Copying agent configuration file '%s' to sandbox (%s)." % (cf_url, pilot_sandbox)
                         logentries.append(Logentry (msg, logger=logger.debug))
-                        sandbox_tgt.copy(cf_url, 'agent_0.cfg')
+                        if shared_filesystem:
+                            sandbox_tgt.copy(cf_url, agent_cfg_name)
 
-                        # close and remove temp file
+                        # Close agent config file
                         os.close(cfg_tmp_handle)
-                        os.unlink(cf_tmp_file)
```
Comment (andre-merzky) [inline src/radical/pilot/controller/pilot_launcher_worker.py]: Why no unlink?

File: src/radical/pilot/controller/pilot_launcher_worker.py [hunk 4]
```diff
@@ -718,8 +726,45 @@ def run(self):
...
+                                    #'%s > %s' % (rp_sdist_path, os.path.basename(rp_sdist_path)),
+                                    '%s > %s' % (rp_sdist_path, os.path.join(jd.working_directory, 'input', os.path.basename(rp_sdist_path))),
+                                    #'%s > %s' % (saga.sdist_path, os.path.basename(saga.sdist_path)),
+                                    '%s > %s' % (saga.sdist_path, os.path.join(jd.working_directory, 'input', os.path.basename(saga.sdist_path))),
+                                    #'%s > %s' % (ru.sdist_path, os.path.basename(ru.sdist_path)),
+                                    '%s > %s' % (ru.sdist_path, os.path.join(jd.working_directory, 'input', os.path.basename(ru.sdist_path)))
+                                ])
+
+                            if agent_config:
+                                jd.file_transfer.append('%s > %s' % (cfg_tmp_file, os.path.join(jd.working_directory, 'input', agent_cfg_name)))
```
Comment (andre-merzky) [inline src/radical/pilot/controller/pilot_launcher_worker.py]: We always have a cfg file.  I think your assumption is that in the default agent_config we won't need it on the other hosts?  Will that always be true?  Why do we then copy the dists?  I am probably missing the point here wrt to unavailable shared FS...

**Comment by andre-merzky** [comment]
This looks overall pretty good to me actually.  Do you consider this a good merge point, or do you think this will need more work before?

**Comment by andre-merzky** [comment]
Mark? :)

**Comment by mturilli** [comment]
I am testing devel already while collecting data for AIMES-Experience. If you are going to merge this into devel, I will add OSG to those tests. It might be a good opportunity.

File: src/radical/pilot/agent/radical-pilot-agent-multicore.py (hunk 1 above)
Comment (marksantcroos) [inline src/radical/pilot/agent/radical-pilot-agent-multicore.py]: So far I've only made MPI tasks optional. If you want to do the same for regular tasks, that needs changes in more places.

Comment (andre-merzky) [inline src/radical/pilot/agent/radical-pilot-agent-multicore.py]: There was a change to `get` in line 4618, that is why I am asking.  But if that is not related, then happy to leave as is...

File: src/radical/pilot/controller/pilot_launcher_worker.py (hunk 4 above)
Comment (marksantcroos) [inline src/radical/pilot/controller/pilot_launcher_worker.py]: > We always have a cfg file.

Good point, changed that.

File: src/radical/pilot/agent/radical-pilot-agent-multicore.py (hunk 1 above)
Comment (marksantcroos) [inline src/radical/pilot/agent/radical-pilot-agent-multicore.py]: It is related, but changing the case for regular tasks here wouldn't make any difference because of other places in the code (both client and agent). So I suggest to leave it indeed.

Comment (andre-merzky) [inline src/radical/pilot/agent/radical-pilot-agent-multicore.py]: ack

**Comment by andre-merzky** [comment]
Thanks! :)
